### ISSUE TITLE
请问您的代码是否保留了L1损失？

### ISSUE THREAD

**Issue by yuanqianguang** [cxliu0/KL-Loss-pytorch#4]
作者您好，您的工作十分有意义，但是我在看KL Loss原文的时候，看其基于caffe2代码的时候，在回归损失应该是包含了smooth L1损失和其提出的KL损失，代码片段如下：

`        loss_bbox = model.net.SmoothL1Loss(
            [
                'bbox_pred', 'bbox_targets', 'bbox_inside_weights',
                'bbox_pred_std_abs' if not cfg.PRED_STD_LOG else 'bbox_pred_std_nexp'
            ],
            'loss_bbox',
            scale=model.GetLossScale()
        )
        
        bbox_pred_std_abs_logw_loss = model.net.SumElements(
                'bbox_pred_std_abs_logwr', 'bbox_pred_std_abs_logw_loss')
        model.net.Scale('bbox_inws_out', 'bbox_inws_out', scale=model.GetLossScale())
        model.net.ReduceMean('bbox_inws_out', 'bbox_inws_outr', axes=[0])
        bbox_pred_std_abs_mulw_loss = model.net.SumElements(
                ['bbox_inws_outr'], 'bbox_pred_std_abs_mulw_loss')                                                          
        loss_gradients = blob_utils.get_loss_gradients(model, [loss_cls, 
            bbox_pred_std_abs_mulw_loss,                                                                                      
            bbox_pred_std_abs_logw_loss                                                                                        
            ] + [loss_bbox])
        model.AddLosses(['loss_cls', 
            'bbox_pred_std_abs_mulw_loss', 'bbox_pred_std_abs_logw_loss'
            ] + ['loss_bbox'])`

但是在您的代码中，我运行了您基于pytorch的代码，发现在回归损失方面仅有KL损失，而没有smooth L1损失，并且代码内容中似乎也确认没有包含smooth L1损失，代码如下

`                losses['loss_bbox'] = (torch.exp(-pos_bbox_var) * l1_loss + 0.5 * pos_bbox_var).sum() / bbox_targets.size(0) * self.loss_kl_coef                
                losses['pos_bbox_var'] = pos_bbox_var
            else:
                losses['loss_bbox'] = bbox_pred.sum() * 0
        return losses`

不知道是否我的理解错误还是没有理解您的代码细节，或者是我没有理解原文作者中的意思。
在原文作者的issue中提到了阻断了KL损失到bbox分支的梯度传递，如果按照使用KL损失替换smooth L1的损失的想法，会使得bbox分支无法得到训练。
但是如果按照同时保留了smooth L1损失和KL损失的想法，在您的KL损失方法的使用代码中我似乎没有找到smooth L1损失的存在，并且没有看到在KL损失中的梯度阻断。
望解惑，谢谢

**Comment by yuanqianguang** [comment]
`bbox_pred_std_abs_logw_loss = model.net.SumElements(
        'bbox_pred_std_abs_logwr', 'bbox_pred_std_abs_logw_loss')
model.net.Scale('bbox_inws_out', 'bbox_inws_out', scale=model.GetLossScale())
model.net.ReduceMean('bbox_inws_out', 'bbox_inws_outr', axes=[0])
bbox_pred_std_abs_mulw_loss = model.net.SumElements(
        ['bbox_inws_outr'], 'bbox_pred_std_abs_mulw_loss')
loss_gradients = blob_utils.get_loss_gradients(model, [loss_cls,
    bbox_pred_std_abs_mulw_loss,
    bbox_pred_std_abs_logw_loss
    ] + [loss_bbox])
model.AddLosses(['loss_cls', 
    'bbox_pred_std_abs_mulw_loss', 'bbox_pred_std_abs_logw_loss'
    ] + ['loss_bbox'])`

`l1_loss = self.loss_bbox(
    pos_bbox_pred,
    bbox_targets[pos_inds.type(torch.bool)],
    bbox_weights[pos_inds.type(torch.bool)],
    reduction_override='none'
)

losses['loss_bbox'] = (torch.exp(-pos_bbox_var) * l1_loss + 0.5 * pos_bbox_var).sum() / bbox_targets.size(0) * self.loss_kl_coef                
losses['pos_bbox_var'] = pos_bbox_var`

抱歉，代码引用如上，格式没有调整好

**Comment by cxliu0** [comment]
1. We use L1 loss instead of Smooth-L1 loss (see [convfc_bbox_head_kl.py#L325](https://github.com/cxliu0/KL-Loss-pytorch/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head_kl.py#L325)), because we found that L1 loss is better than Smooth-L1 loss. 
2. We implement KL loss following Eq. 10 in the paper, i.e., KL loss is implemented as ```e^-alpha * L1_Loss + 1/2 * alpha```.

**Comment by yuanqianguang** [comment]
> 1. We use L1 loss instead of Smooth-L1 loss (see [convfc_bbox_head_kl.py#L325](https://github.com/cxliu0/KL-Loss-pytorch/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head_kl.py#L325)), because we found that L1 loss is better than Smooth-L1 loss.
> 2. We implement KL loss following Eq. 10 in the paper, i.e., KL loss is implemented as `e^-alpha * L1_Loss + 1/2 * alpha`.

Thanks for your reply, but it seems that there is only KL Loss in regression loss without the L1 Loss which is just used to compute the first term of KL Loss.
and it seems the gradient of sigma is backprop to the bbox branch in your code because I can't find the operation of stopping gradient from sigma branch to bbox branch.
Wish for your more reply, thx.

**Comment by cxliu0** [comment]
First, KL loss already contains L1 regression loss (see [convfc_bbox_head_kl.py#L332](https://github.com/cxliu0/KL-Loss-pytorch/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head_kl.py#L332)), where ```l1_loss``` is L1 regression loss and ```pos_bbox_var``` corresponds to ```alpha``` in Eq. 10.

Second, we do not stop the gradient of sigma, as we did not observe improvement when adopting such an operation. You may try it if you like.

**Comment by yuanqianguang** [comment]
emmm, I know the second, but the first explanation......
Maybe I know what you mean. The `KL Loss + L1 Loss` is equal to `KL Loss` because the alpha of `e^-alpha * L1_Loss` can learn the additional effect of `1` from the L1 Loss? that's right?
Or you mean that the performance of using the KL Loss to replace L1 Loss is better than using KL Loss + L1 Loss?
I‘m confused because it seems there are `KL Loss + L1 Loss` in the original KL Loss code.
[https://github.com/ethanhe42/KL-Loss/blob/master/detectron/modeling/fast_rcnn_heads.py#L173](url)

**Comment by cxliu0** [comment]
According to my knowledge, I believe you may misunderstand the original KL loss. 

In [https://github.com/ethanhe42/KL-Loss/blob/master/detectron/modeling/fast_rcnn_heads.py#L173](https://github.com/ethanhe42/KL-Loss/blob/master/detectron/modeling/fast_rcnn_heads.py#L173):
- ```loss_bbox``` is the modulated Smooth-L1 loss, i.e., ```e^-alpha * Smooth-L1``` (see [https://github.com/ethanhe42/KL-Loss/blob/master/detectron/modeling/fast_rcnn_heads.py#L157](https://github.com/ethanhe42/KL-Loss/blob/master/detectron/modeling/fast_rcnn_heads.py#L157)), which corresponds to the first term of Eq. 9 and Eq. 10.
- ```bbox_pred_std_abs_mulw_loss``` is also the modulated Smooth-L1 loss (with the same value as ```loss_bbox```), but its gradient is stopped. It does not contribute to model training.
- ```bbox_pred_std_abs_logw_loss``` corresponds to ```1/2 * alpha``` in Eq. 9 and Eq. 10.

Therefore, the final regression loss that with effective gradients is actually ```e^-alpha * Smooth-L1 + 1/2 * alpha```, not ```KL loss + Smooth-L1 loss```. 

Perhaps the authors just use ```bbox_pred_std_abs_mulw_loss``` to validate that their loss computation is correct, or to ensure that the final regression loss is larger than 0.

**Comment by yuanqianguang** [comment]
Thank you for your fast and patient response  :) 
I'm not familiar with caffe and detectron so that maybe I misunderstand something from the original code of KL Loss paper.
So, The whole regression loss is just KL Loss which supervise the coordinate prediction and standard prediction, is that right?

**Comment by cxliu0** [comment]
The regression loss is KL loss, which corresponds to Eq. 9 and Eq. 10 in the paper.

**Comment by yuanqianguang** [comment]
OK, Thanks so much~ ^-^
